src/backend/distributed/executor/insert_select_executor.c

```diff
@@ -0,0 +1,157 @@
+ */
+
+#include "postgres.h"
+
+#include "distributed/insert_select_executor.h"
+#include "distributed/insert_select_planner.h"
+#include "distributed/multi_copy.h"
+#include "distributed/multi_executor.h"
+#include "distributed/multi_physical_planner.h"
+#include "distributed/multi_planner.h"
+#include "distributed/transaction_management.h"
+#include "executor/executor.h"
+#include "nodes/execnodes.h"
+#include "nodes/makefuncs.h"
+#include "nodes/nodeFuncs.h"
+#include "nodes/parsenodes.h"
+#include "nodes/plannodes.h"
+#include "parser/parse_coerce.h"
+#include "parser/parsetree.h"
+#include "tcop/pquery.h"
+#include "tcop/tcopprot.h"
+#include "utils/lsyscache.h"
+#include "utils/portal.h"
+#include "utils/snapmgr.h"
+
+
+static void ExecuteSelectIntoRelation(Oid targetRelationId, List *insertTargetList,
+									  Query *selectQuery, EState *executorState);
+static void ExecuteIntoDestReceiver(Query *query, ParamListInfo params,
+									DestReceiver *dest);
+
+
+/*
+ * CoordinatorInsertSelectExecScan executes an INSERT INTO distributed_table
+ * SELECT .. query by setting up a DestReceiver that copies tuples into the
+ * distributed table and then executing the SELECT query using that DestReceiver
+ * as the tuple destination.
+ */
+TupleTableSlot *
+CoordinatorInsertSelectExecScan(CustomScanState *node)
+{
+	CitusScanState *scanState = (CitusScanState *) node;
+	TupleTableSlot *resultSlot = NULL;
+
+	if (!scanState->finishedRemoteScan)
+	{
+		EState *executorState = scanState->customScanState.ss.ps.state;
+		MultiPlan *multiPlan = scanState->multiPlan;
+		Query *selectQuery = multiPlan->insertSelectSubquery;
+		List *insertTargetList = multiPlan->insertTargetList;
+		Oid targetRelationId = multiPlan->targetRelationId;
+
+		ereport(DEBUG1, (errmsg("Collecting INSERT ... SELECT results on coordinator")));
+
+		ExecuteSelectIntoRelation(targetRelationId, insertTargetList, selectQuery,
+								  executorState);
+
+		scanState->finishedRemoteScan = true;
+	}
+
+	resultSlot = ReturnTupleFromTuplestore(scanState);
+
+	return resultSlot;
+}
+
+
+/*
+ * ExecuteSelectIntoRelation executes given SELECT query and inserts the
+ * results into the target relation, which is assumed to be a distributed
+ * table.
+ */
+static void
+ExecuteSelectIntoRelation(Oid targetRelationId, List *insertTargetList,
+						  Query *selectQuery, EState *executorState)
+{
+	ParamListInfo paramListInfo = executorState->es_param_list_info;
+
+	ListCell *insertTargetCell = NULL;
+	List *columnNameList = NIL;
+	bool stopOnFailure = false;
+	char partitionMethod = 0;
+
+	CitusCopyDestReceiver *copyDest = NULL;
+
+	BeginOrContinueCoordinatedTransaction();
+
+	partitionMethod = PartitionMethod(targetRelationId);
+	if (partitionMethod == DISTRIBUTE_BY_NONE)
+	{
+		stopOnFailure = true;
+	}
+
+	/* build the list of column names for the COPY statement */
+	foreach(insertTargetCell, insertTargetList)
+	{
```

Comment:
While I was randomly testing the changes, I hit an issue which is not straightforward to understand (at least for me). So, I'm posting here in case you have any ideas.

```SQL

CREATE TABLE agg_results (user_id int, value_1_agg int, value_2_agg int, value_3_agg float, value_4_agg bigint, agg_time timestamp);
SELECT create_distributed_table('agg_results', 'user_id');
CREATE TABLE users_table_local (user_id int, time timestamp, value_1 int, value_2 int, value_3 float, value_4 bigint);

INSERT INTO users_table_local (user_id) VALUES (1);

INSERT INTO agg_results  (value_4_agg, user_id)
(
	SELECT value_4::float, user_id FROM users_table_local
)
UNION
(
	SELECT value_4, user_id FROM users_table_local
);
ERROR:  the partition column of table public.agg_results should have a value

```